File: backend/controllers/spotifyController.ts
import { Request, Response } from "express";
import * as spotifyService from "../services/spotifyService";
import "../loadEnv.js";
import { Track } from "../types.js";
import dotenv from "dotenv";
import prisma from 'prisma'
import axios from 'axios'

dotenv.config();

const CLIENT_ID = process.env.SPOTIFY_CLIENT_ID;
const CLIENT_SECRET = process.env.SPOTIFY_CLIENT_SECRET;
const REDIRECT_URI = process.env.SPOTIFY_REDIRECT_URI!; // Backend Redirect URI
const SPOTIFY_AUTH_URL = "https://accounts.spotify.com/authorize";
const SPOTIFY_TOKEN_URL = "https://accounts.spotify.com/api/token";
const SCOPES = [
  "playlist-read-private",
  "playlist-read-collaborative",
  "playlist-modify-private",
  "playlist-modify-public",
];

export const loginWithSpotify = (req: Request, res: Response) => {
  console.log(CLIENT_ID, CLIENT_SECRET, REDIRECT_URI)
  const authURL = `${SPOTIFY_AUTH_URL}?client_id=${CLIENT_ID}&response_type=code&redirect_uri=${encodeURIComponent(
    REDIRECT_URI
  )}&scope=${encodeURIComponent(SCOPES.join(" "))}`;

  res.redirect(authURL);
};

export const handleSpotifyCallback = async (req: Request, res: Response) => {
  try {
    const { code } = req.query;
    if (!code) return res.status(400).json({ message: "Authorization code is missing" });

    // Request access token from Spotify
    const tokenResponse = await axios.post(
      SPOTIFY_TOKEN_URL,
      new URLSearchParams({
        grant_type: "authorization_code",
        code: code as string,
        redirect_uri: REDIRECT_URI,
        client_id: CLIENT_ID,
        client_secret: CLIENT_SECRET
      }),
      { headers: { "Content-Type": "application/x-www-form-urlencoded" } }
    );


    const { access_token, refresh_token, expires_in } = tokenResponse.data;

    // Fetch user's Spotify ID using the access token
    const userProfile = await axios.get("https://api.spotify.com/v1/me", {
      headers: { Authorization: `Bearer ${access_token}` }
    });

    const spotifyId = userProfile.data.id;

    // Redirect to frontend with a success flag (store session in frontend)
    // res.redirect(`${process.env.FRONTEND_URL}/tunescout/dashboard?sucess=True`);
    res.redirect(`${process.env.FRONTEND_URL}/tunescout/dashboard`);
  } catch (error) {
    console.error("Error exchanging code for token:", error.response?.data || error.message);
    res.status(500).json({ message: "Failed to authenticate with Spotify" });
  }
};

export const refreshSpotifyToken = async (req: Request, res: Response) => {
  try {
    const { email } = req.body;
    if (!email) return res.status(400).json({ message: "User email is required" });

    const user = await prisma.user.findUnique({ where: { email } });
    if (!user || !user.spotifyRefreshToken)
      return res.status(401).json({ message: "User is not authenticated with Spotify" });

    // Request new access token
    const tokenResponse = await axios.post(
      SPOTIFY_TOKEN_URL,
      new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: user.spotifyRefreshToken,
        client_id: CLIENT_ID,
        client_secret: CLIENT_SECRET
      }),
      { headers: { "Content-Type": "application/x-www-form-urlencoded" } }
    );

    const { access_token } = tokenResponse.data;

    // Update access token in database
    await prisma.user.update({
      where: { email },
      data: { spotifyAccessToken: access_token }
    });

    res.status(200).json({ access_token });
  } catch (error) {
    console.error("Error refreshing token:", error.response?.data || error.message);
    res.status(500).json({ message: "Failed to refresh Spotify token" });
  }
};

// ✅ Get user playlists
export const getUserPlaylists = async (req: Request, res: Response) => {
  try {
    const { access_token } = req.body;
    if (!access_token)
      return res.status(400).json({ message: "Access token is required" });

    const playlists = await spotifyService.getUserPlaylists(access_token);
    res.status(200).json(playlists);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// ✅ Generate playlist (fetch similar tracks, convert to Spotify format)
export const generatePlaylist = async (req: Request, res: Response) => {
  try {
    const { playlist_id, access_token, number_of_songs } = req.body;
    if (!access_token || !playlist_id)
      throw new Error("Missing required fields");

    const originalTracks = await spotifyService.fetchAllPlaylistTracks(
      playlist_id,
      access_token
    );
    if (originalTracks.length < 10)
      throw new Error("Playlist does not have enough tracks");

    const similarTracks: Track[] =
      await spotifyService.fetchEnoughSimilarTracks(
        originalTracks,
        number_of_songs
      );
    const removeTracks = getRandomValuesFromList(
      similarTracks,
      number_of_songs
    );
    const finalTracks = await convertToSpotify(removeTracks, access_token);

    res.status(200).json(finalTracks);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// ✅ Create a new Spotify playlist & add tracks
export const createPlaylist = async (req: Request, res: Response) => {
  try {
    const { access_token, tracks, name } = req.body;
    if (!access_token || !tracks || !name)
      throw new Error("Missing required fields");

    const playlistData = await spotifyService.createPlaylist(
      name,
      access_token
    );
    await spotifyService.addTracksToPlaylist(
      playlistData.id,
      tracks.map((track) => track.uri),
      access_token
    );

    res.status(200).json({
      message: "Playlist created successfully",
      playlistId: playlistData.id,
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// ✅ Helper functions (moved from main code)
const convertToSpotify = async (similarTracks, access_token) => {
  const spotifyTrackPromises = similarTracks.map(async (track) => {
    return await spotifyService.fetchSpotifyTrackId(
      track.name,
      track.artist.name,
      access_token
    );
  });

  const resolvedTracks = await Promise.all(spotifyTrackPromises);
  return resolvedTracks.filter((track) => track !== null);
};

const getRandomValuesFromList = (list, x) => {
  return list.length <= x
    ? list
    : list.sort(() => 0.5 - Math.random()).slice(0, x);
};

File: fastapi-backend/app/controllers/spotify_controller.py
import httpx
from fastapi import Body, Header, HTTPException
from fastapi.responses import JSONResponse

from app.schemas.playlists import AuthCodeBody, GeneratePlaylistRequest
from app.services import spotify_service
import logging


async def check_authorization(authorization: str):
    if not authorization:
        raise HTTPException(status_code=400, detail="Authorization header is required")
    if not authorization.startswith("Bearer "):
        raise HTTPException(status_code=400, detail="Invalid Authorization format")

    token = authorization.removeprefix("Bearer ").strip()
    return token


async def get_playlists(authorization: str = Header(...)):
    token = await check_authorization(authorization)
    try:
        playlists = await spotify_service.get_playlists(token)
        return JSONResponse(content=playlists, status_code=200)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


async def generate_playlist(
    authorization: str,
    request: GeneratePlaylistRequest,  # noqa: B008
):
    """Create an API endpoint to generate a playlist from Spotify.

    Args:
        authorization (str): Spotify authorization token.
        request (GeneratePlaylistRequest): Request body containing the playlist ID and
            number of songs.

    """
    token = await check_authorization(authorization)
    try:
        logger = logging.getLogger(__name__)
        logger.warning("I got here")
        original_tracks = await spotify_service.fetch_all_tracks(
            token, request.playlist_id
        )
        if len(original_tracks) <= 10:
            raise HTTPException(
                status_code=500, detail="Playlist does not have enough tracks."
            )

        similar_tracks = await spotify_service.fetch_similar_tracks(
            original_tracks, request.number_of_songs
        )
        spotify_tracks = await spotify_service.convert_to_spotify(token, similar_tracks)
        return JSONResponse(content=spotify_tracks, status_code=200)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


async def create_playlist(
    authorization: str = Header(...),
    request: GeneratePlaylistRequest = Body(...),  # noqa: B008
):
    pass


async def callback(request: AuthCodeBody):
    """Complete PKCE flow and get access token from Spotify.

    Args:
        request (AuthCodeBody): AuthCodeBody schema containing the code, redirect_uri,
        client_id, and code_verifier.

    Returns:
        response.json: JSON response from Spotify containing the access token.

    """
    async with httpx.AsyncClient() as client:
        base_url = "https://accounts.spotify.com/api/token"
        params = {
            "client_id": request.client_id,
            "grant_type": "authorization_code",
            "code": request.code,
            "redirect_uri": request.redirect_uri,
            "code_verifier": request.code_verifier,
        }

        response = await client.post(
            base_url,
            data=params,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

        return response.json()
